### PR TITLE
added blockhash input to claimtrie RPC methods plus tests

### PR DESCRIPTION
**Pull request by BrannonKing**
I'm marking this as WIP as I'm unsure on the approach and looking for feedback. You'll notice that `getclaimsforname` is done differently (more robustly) than the others. The others could be done like that one but not vice-versa, but I'm less sure on the consequences of that approach and the completeness of what's there.

### CODE REVIEW

**Comment by BrannonKing** [comment]
Apparently I need to add those minconf parameters to the client.cpp file for automatic CLI conversion to int.

File: src/claimtrie.cpp [hunk 1]
```diff
@@ -2554,10 +2557,75 @@ uint256 CClaimTrieCache::getLeafHashForProof(const std::string& currentPosition,
...
+        cachedNode = cache.find(str);
+        if (cachedNode != cache.end())
+            recursiveFlatten(str, cachedNode->second, nodes);
+        else
+            recursiveFlatten(str, it->second, nodes);
+    }
+}
+
+
+std::vector<namedNodeType> CClaimTrieCache::flattenTrie() const
```
Comment (BrannonKing) [inline src/claimtrie.cpp]: This method existed for CClaimTrie but not for the "cache".

File: src/claimtrie.cpp [hunk 2]
```diff
@@ -2554,10 +2557,75 @@ uint256 CClaimTrieCache::getLeafHashForProof(const std::string& currentPosition,
...
+        if (cachedNode != cache.end())
+            recursiveFlatten(str, cachedNode->second, nodes);
+        else
+            recursiveFlatten(str, it->second, nodes);
+    }
+
+    return nodes;
+}
+
+bool CClaimTrieCache::getInfoForName(const std::string& name, CClaimValue& toBeFilled) const
```
Comment (BrannonKing) [inline src/claimtrie.cpp]: this method also existed for CClaimTrie but not for the cache; it was mostly duplicated here.

File: src/claimtrie.cpp [hunk 3]
```diff
@@ -2575,9 +2643,10 @@ CClaimTrieProof CClaimTrieCache::getProofForName(const std::string& name) const
         uint256 valueHash;
         if (fNodeHasValue)
         {
-            int nHeightOfLastTakeover;
-            assert(getLastTakeoverForName(currentPosition, nHeightOfLastTakeover));
-            valueHash = getValueHash(claim.outPoint, nHeightOfLastTakeover);
+            int nHeightOfLastTakeover2;
+            if (!getLastTakeoverForName(currentPosition, nHeightOfLastTakeover2))
+                return false;
```
Comment (BrannonKing) [inline src/claimtrie.cpp]: I was making this follow the pattern of `getInfoForName`.

File: src/rpc/claimtrie.cpp [hunk 4]
```diff
@@ -46,55 +48,54 @@ UniValue getclaimsintrie(const UniValue& params, bool fHelp)
...
-    CCoinsViewCache view(pcoinsTip);
-    std::vector<namedNodeType> nodes = pclaimTrie->flattenTrie();
+    int maxHeight = chainActive.Height();
+    if (params.size() > 0)
+        maxHeight -= params[0].get_int() - 1;
 
+    CCoinsViewCache coins(pcoinsTip);
+    CClaimTrieCache claimTrie(pclaimTrie);
+    if (maxHeight < chainActive.Height())
+        if (!RollBackTo(coins, claimTrie, chainActive[maxHeight]))
```
Comment (BrannonKing) [inline src/rpc/claimtrie.cpp]: This method uses the approach of rolling back the caches to put us in a state where we were some number of blocks back. The caches are never flushed.

File: src/rpc/claimtrie.cpp [hunk 5]
```diff
@@ -197,19 +209,32 @@ UniValue getvalueforname(const UniValue& params, bool fHelp)
...
+    if (!getValueForClaim(coins, claim.outPoint, sValue))
         return ret;
+
     ret.push_back(Pair("value", sValue));
     ret.push_back(Pair("claimId", claim.claimId.GetHex()));
     ret.push_back(Pair("txid", claim.outPoint.hash.GetHex()));
     ret.push_back(Pair("n", (int)claim.outPoint.n));
     ret.push_back(Pair("amount", claim.nAmount));
-    ret.push_back(Pair("effective amount", pclaimTrie->getEffectiveAmountForClaim(name, claim.claimId))); 
+    ret.push_back(Pair("effective amount", pclaimTrie->getEffectiveAmountForClaim(name, claim.claimId, maxHeight)));
```
Comment (BrannonKing) [inline src/rpc/claimtrie.cpp]: This particular line is not using data from the caches. It's hitting the live claimTrie itself. It appears that passing in the maxHeight would give the same result as rolling back the main trie.

File: src/rpc/claimtrie.cpp [hunk 6]
```diff
@@ -323,6 +349,42 @@ UniValue getclaimsforname(const UniValue& params, bool fHelp)
...
+    } BOOST_SCOPE_EXIT_END
+
+    if (params.size() > 1)
+    {
+        int minconf = params[1].get_int();
+        while(--minconf > 0) {
+            indexes.push_back(chainActive.Tip());
+            CBlock block;
+            ReadBlockFromDisk(block, indexes.back(), Params().GetConsensus());
+            DisconnectBlock(block, state, indexes.back(), coins, claims);
```
Comment (BrannonKing) [inline src/rpc/claimtrie.cpp]: In this method we are actually rolling back the real blocks -- not just the caches. We hold the locks and disconnect some number of tips and then rely on our scope-exit handler to put everything back before we leave. I think it's a horrible approach, really, but didn't see a good way to do it otherwise in the current codebase. We lock the entire system for handling this method. That's so wrong. I think Roy's approach of storing all claimtrie (nodes that have ever existed) in LevelDB is much better -- the old nodes would still be there and you could just start a search at a certain time period -- we wouldn't ever have to lock; we could run numerous calls concurrently.

File: src/rpc/claimtrie.cpp [hunk 7]
```diff
@@ -323,6 +349,42 @@ UniValue getclaimsforname(const UniValue& params, bool fHelp)
...
+            // assuming we don't need to resurrect the memory pool entries for this RPC call
+        }
+        coins.Flush(); // need flushes on every iteration?
+        claims.flush();
+    }
+
     claimsForNameType claimsForName = pclaimTrie->getClaimsForName(name);
```
Comment (BrannonKing) [inline src/rpc/claimtrie.cpp]: There was no equivalent of `getClaimsForName` in the cache object, and it was not obvious how to achieve this method's "support" outputs using the data in the cache.

File: src/claimtrie.cpp (hunk 1 above)
Comment (lbrynaut) [reply src/claimtrie.cpp]: What is this supposed to do?  It doesn't make sense to me.

File: src/rpc/claimtrie.cpp (hunk 6 above)
Comment (lbrynaut) [reply src/rpc/claimtrie.cpp]: All (current) claimtrie nodes are stored in leveldb.  Storing all changes over time (while nice for analysis) isn't actually that practical of a feature to have.  Eeven concurrent performance on lookups isn't needed for just about anything practical (other than analysis), where I'd argue a relational DB is just as well suited, if not better.  Horrible or not, all persistent state changes of a blockchain occur via connecting and disconnecting blocks.

Now is all this needed for this particular issue?  I'm not sure yet, haven't thought about it enough.

File: src/claimtrie.cpp (hunk 1 above)
Comment (bvbfan) [reply src/claimtrie.cpp]: We need ability read cached claim, it makes sense when we use rollback feature.

File: src/claimtrie.cpp [hunk 8]
```diff
@@ -2574,10 +2562,125 @@ uint256 CClaimTrieCache::getLeafHashForProof(const std::string& currentPosition,
...
+
+bool CClaimTrieCache::getInfoForName(const std::string& name, CClaimValue& claim) const
+{
+    if (dirty())
+        getMerkleHash();
+
+    CClaimTrieNode* current = &(base->root);
+    nodeCacheType::const_iterator cachedNode;
+
+    for (std::string::const_iterator itName = name.begin(); current; ++itName) {
```
Comment (BrannonKing) [inline src/claimtrie.cpp]: use `getNodeForName` here?

Comment (bvbfan) [reply src/claimtrie.cpp]: I've thinking about that, but getNodeForName fallback to base (CClaimTrie) when it's not found in cache.

**Comment by bvbfan** [comment]
I plan to standardized JSON objects names, between function they have different names which is ugly and hard to parse in a script.
claimId
txId
txn
value
presentInTrie (in claim trie)
effectiveAmount
validAtHeight
and so on, first word starts with lower letter, second word, if it presents, starts with upper. Most help does not match returned JSON object, Any other suggestions?

**Comment by BrannonKing** [comment]
> I plan to standardized JSON objects names...

We need to do that as a separate pull request. That's a breaking change, whereas adding additional outputs and taking new optional inputs are not breaking changes. I agree that the current mismatches between methods are silly. Do you know where each RPC method is used in the LBRY stack? We should locate each call and write up an epic story that includes updating each method that is necessary.

**Comment by bvbfan** [comment]
> We need to do that as a separate pull request. That's a breaking change, whereas adding additional outputs and taking new optional inputs are not breaking changes. I agree that the current mismatches between methods are silly. 

I understand, you're right.

>Do you know where each RPC method is used in the LBRY stack? We should locate each call and write up an epic story that includes updating each method that is necessary.

lbry.go, i think, but @shyba knows better.

**Comment by bvbfan** [comment]
Rebase to be ready for merge.

File: src/claimtrie.cpp [hunk 9]
```diff
@@ -527,39 +527,26 @@ claimsForNameType CClaimTrie::getClaimsForName(const std::string& name) const
...
+        if (it->claimId == claimId && it->nValidAtHeight < nCurrentHeight) {
             effectiveAmount += it->nAmount;
-            claim_found = true;
+            for (std::vector<CSupportValue>::const_iterator it = claims.supports.begin(); it != claims.supports.end(); ++it) {
+                if (it->supportedClaimId == claimId && it->nValidAtHeight < nCurrentHeight) {
+                    effectiveAmount += it->nAmount;
+                    if (supports) supports->push_back(*it);
+                }
+            }
             break;
```
Comment (kaykurokawa) [inline src/claimtrie.cpp]: What are all these changes to getEffectiveAmoutnForClaim for? Seems like an unrelated change?

File: src/rpc/claimtrie.cpp [hunk 10]
```diff
@@ -22,143 +24,177 @@ uint160 ParseClaimtrieId(const UniValue& v, const std::string& strName)
...
+    if (mapBlockIndex.count(blockHash) == 0)
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+
+    CBlockIndex* pblockIndex = mapBlockIndex[blockHash];
+    if (!chainActive.Contains(pblockIndex))
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Block not in main chain");
+
+    if (chainActive.Tip()->nHeight > (pblockIndex->nHeight + MAX_RPC_BLOCK_DECREMENTS))
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Block too deep to regenerate it");
+
```
Comment (kaykurokawa) [inline src/rpc/claimtrie.cpp]: Instead of "Block too deep to regenerate it" , how about just "Block is too deep"

File: src/rpc/claimtrie.cpp [hunk 11]
```diff
@@ -22,143 +24,177 @@ uint160 ParseClaimtrieId(const UniValue& v, const std::string& strName)
...
             "Return all claims in the name trie.\n"
             "Arguments:\n"
-            "None\n"
+            "1. \"blockhash\"     (string, optional) the hash of the block\n"
+            "                                           which is the active\n"
+            "                                           in the chain. If\n"
+            "                                           none is given, \n"
+            "                                           the latest block\n"
+            "                                           will be used.\n"
             "Result: \n"
```
Comment (kaykurokawa) [inline src/rpc/claimtrie.cpp]: instead of "the hash of the block which is the active in the chain. If none is given, the latest block will be used", how about:

"get claims in the trie at the block specified by this block hash. If none is given, the latest active block will be used"

File: src/rpc/claimtrie.cpp [hunk 12]
```diff
@@ -22,143 +24,177 @@ uint160 ParseClaimtrieId(const UniValue& v, const std::string& strName)
...
             "Return the entire name trie.\n"
             "Arguments:\n"
-            "None\n"
+            "1. \"blockhash\"   (string, optional) the hash of the block\n"
+            "                                         which is the active\n"
+            "                                         in the chain. If\n"
+            "                                         none is given, \n"
+            "                                         the latest block\n"
+            "                                         will be used.\n"
             "Result: \n"
```
Comment (kaykurokawa) [inline src/rpc/claimtrie.cpp]: how about:
"get the claim trie at the block specified by this block hash. If none is given, the latest active block will be used"

File: src/rpc/claimtrie.cpp [hunk 13]
```diff
@@ -180,118 +216,112 @@ bool getValueForClaim(const COutPoint& out, std::string& sValue)
...
             "Arguments:\n"
-            "1. \"name\"             (string) the name to look up\n"
+            "1. \"name\"          (string) the name to look up\n"
+            "2. \"blockhash\"     (string, optional) the hash of the block\n"
+            "                                           which is the active\n"
+            "                                           in the chain. If\n"
+            "                                           none is given, \n"
+            "                                           the latest block\n"
+            "                                           will be used.\n"
             "Result: \n"
```
Comment (kaykurokawa) [inline src/rpc/claimtrie.cpp]: how about "get the value associated with the name at the block specified by this block hash. If none is given, the latest active block will be used"

File: src/rpc/claimtrie.cpp [hunk 14]
```diff
@@ -180,118 +216,112 @@ bool getValueForClaim(const COutPoint& out, std::string& sValue)
...
             "Arguments: \n"
-            "1.  \"name\"       (string) the name for which to get claims and supports\n"
+            "1.  \"name\"         (string) the name for which to get claims and supports\n"
+            "2.  \"blockhash\"    (string, optional) the hash of the block\n"
+            "                                           which is the active\n"
+            "                                           in the chain. If\n"
+            "                                           none is given, \n"
+            "                                           the latest block\n"
+            "                                           will be used.\n"
             "Result:\n"
```
Comment (kaykurokawa) [inline src/rpc/claimtrie.cpp]: how about, "get claims for name at the block specified by this block hash. If none is given, the latest active block will be used"

**Comment by kaykurokawa** [comment]
I added a commit for some unit tests for the new CClaimtrieCache functions in claimtriecache_tests.cpp, by no means extensive but I think it'd be good to have some coverage on those. 

https://github.com/lbryio/lbrycrd/pull/189/commits/0159a57765154ca2a09df4fc3845546eb3908a00

File: src/claimtrie.cpp (hunk 9 above)
Comment (bvbfan) [reply src/claimtrie.cpp]: When I made a getEffectiveAmountForClaim in CClaimTrieCache notice it has 2 functions in here, so i just want to match them in the 2 classes. I prefer pointer to indicate it's optional or not. In plus now it has overloaded function that takes claimsForNameType if you performed getClaimsForName before to prevent double look-up. Otherwise we end-up with 4 functions
```
getEffectiveAmountForClaim(const std::string& name, uint160 claimId)
getEffectiveAmountForClaim(const claimsForNameType& claims, uint160 claimId)
getEffectiveAmountForClaimWithSupports(const std::string& name, uint160 claimId)
getEffectiveAmountForClaimWithSupports(const claimsForNameType& claims, uint160 claimId)
```

File: src/rpc/claimtrie.cpp (hunk 11 above)
Comment (BrannonKing) [reply src/rpc/claimtrie.cpp]: getnameproof says "Block too deep to generate proof"

**Comment by kaykurokawa** [comment]
Fixed formatting on the test I added.
Also noting that this will affect calls in https://github.com/lbryio/lbryumx for lbryumx/session.py , this should be the only place where this PR will affect. I will open issue there

**Comment by BrannonKing** [comment]
How will libryumx be affected? The intent was that all new parameters here were optional.

**Comment by bvbfan** [comment]
Update help strings.

File: src/rpc/claimtrie.cpp (hunk 10 above)
Comment (BrannonKing) [reply src/rpc/claimtrie.cpp]: getnameproof says "Block too deep to generate proof"

**Comment by BrannonKing** [comment]
I was just doing some testing by hand and seeing a problem:
1. start up a clean regtest daemon (after removing ~/.lbrycrd/regtest)
2. generate 200 blocks
3. claim a new name
4. generate 1 block
5. get the block hash for block 190
6. run getclaimsintrie passing in the hash for 190
7. the results should be empty but they aren't

**Comment by kaykurokawa** [comment]
Noting here that RPC command getclaimbyid does not have a blockhash argument in this PR . 

I think it might be a bit involved to implement .. since we'll have to redo the getClaimById function in CClaimTrie ( it relies on leveldb and a cache, but the cache I think does not support decrementing multiple blocks at a time, just one block at at time) . We could probably open a new PR for this instead of  doing this here.
